docker/resource_docker_container_funcs.go

```diff
@@ -682,7 +682,49 @@ func resourceDockerContainerRead(d *schema.ResourceData, meta interface{}) error
 }
 
 func resourceDockerContainerUpdate(d *schema.ResourceData, meta interface{}) error {
-	// TODO call resourceDockerContainerRead here
+	attrs := []string{
+		"restart", "max_retry_count", "cpu_shares", "memory", "cpu_set", "memory_swap",
+	}
+	for _, attr := range attrs {
+		if d.HasChange(attr) {
+
+			// TODO update ulimits
+			// Updating ulimits seems not to work well.
```

Comment:
What exactly does not work well? Can you be more specific? Let's merge those 2 comments (L693) into one. What do you think?

Reply:
> What exactly does not work well?

It succeeds to run `terraform apply` but actually `ulimit` aren't changed.

```hcl
provider "docker" {
}

resource "docker_image" "alpine" {
  name = "alpine:3.10.3"
}

resource "docker_container" "foo" {
  image      = docker_image.alpine.latest
  name       = "foo"
  entrypoint = ["tail"]
  command    = ["-f", "/dev/null"]
  ulimit {
    name = "nofile"
    hard = 262144
    soft = 200000
  }
}
```

Create a container.

```
$ terraform apply
```

And update `soft` from `200000` to `100000` and run `terraform apply` again.

```
$ terraform apply
docker_image.alpine: Refreshing state... [id=sha256:965ea09ff2ebd2b9eeec88cd822ce156f6674c7e99be082c7efac3c62f3ff652alpine:3.10.3]
docker_container.foo: Refreshing state... [id=c22859d2be1768bfe200075a1351c3e8364925cc3218f583a2e04ec379ab63bf]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # docker_container.foo will be updated in-place
  ~ resource "docker_container" "foo" {
        attach            = false
        command           = [
            "-f",
            "/dev/null",
        ]
        cpu_shares        = 0
        dns               = []
        dns_opts          = []
        dns_search        = []
        entrypoint        = [
            "tail",
        ]
        env               = [
            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
        ]
        gateway           = "172.17.0.1"
        group_add         = []
        hostname          = "c22859d2be17"
        id                = "c22859d2be1768bfe200075a1351c3e8364925cc3218f583a2e04ec379ab63bf"
        image             = "sha256:965ea09ff2ebd2b9eeec88cd822ce156f6674c7e99be082c7efac3c62f3ff652"
        ip_address        = "172.17.0.3"
        ip_prefix_length  = 16
        ipc_mode          = "private"
        links             = []
        log_driver        = "json-file"
        log_opts          = {}
        logs              = false
        max_retry_count   = 0
        memory            = 0
        memory_swap       = 0
        must_run          = true
        name              = "foo"
        network_data      = [
            {
                gateway          = "172.17.0.1"
                ip_address       = "172.17.0.3"
                ip_prefix_length = 16
                network_name     = "bridge"
            },
        ]
        network_mode      = "default"
        privileged        = false
        publish_all_ports = false
        read_only         = false
        restart           = "no"
        rm                = false
        shm_size          = 64
        start             = true
        sysctls           = {}
        tmpfs             = {}

      + ulimit {
          + hard = 262144
          + name = "nofile"
          + soft = 100000
        }
      - ulimit {
          - hard = 262144 -> null
          - name = "nofile" -> null
          - soft = 200000 -> null
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

docker_container.foo: Modifying... [id=c22859d2be1768bfe200075a1351c3e8364925cc3218f583a2e04ec379ab63bf]
docker_container.foo: Modifications complete after 0s [id=c22859d2be1768bfe200075a1351c3e8364925cc3218f583a2e04ec379ab63bf]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

It succeeds to run `terraform apply`, but actually `ulimit` aren't changed.

```
$ terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

docker_image.alpine: Refreshing state... [id=sha256:965ea09ff2ebd2b9eeec88cd822ce156f6674c7e99be082c7efac3c62f3ff652alpine:3.10.3]
docker_container.foo: Refreshing state... [id=c22859d2be1768bfe200075a1351c3e8364925cc3218f583a2e04ec379ab63bf]

------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # docker_container.foo will be updated in-place
  ~ resource "docker_container" "foo" {
        attach            = false
        command           = [
            "-f",
            "/dev/null",
        ]
        cpu_shares        = 0
        dns               = []
        dns_opts          = []
        dns_search        = []
        entrypoint        = [
            "tail",
        ]
        env               = [
            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
        ]
        gateway           = "172.17.0.1"
        group_add         = []
        hostname          = "c22859d2be17"
        id                = "c22859d2be1768bfe200075a1351c3e8364925cc3218f583a2e04ec379ab63bf"
        image             = "sha256:965ea09ff2ebd2b9eeec88cd822ce156f6674c7e99be082c7efac3c62f3ff652"
        ip_address        = "172.17.0.3"
        ip_prefix_length  = 16
        ipc_mode          = "private"
        links             = []
        log_driver        = "json-file"
        log_opts          = {}
        logs              = false
        max_retry_count   = 0
        memory            = 0
        memory_swap       = 0
        must_run          = true
        name              = "foo"
        network_data      = [
            {
                gateway          = "172.17.0.1"
                ip_address       = "172.17.0.3"
                ip_prefix_length = 16
                network_name     = "bridge"
            },
        ]
        network_mode      = "default"
        privileged        = false
        publish_all_ports = false
        read_only         = false
        restart           = "no"
        rm                = false
        shm_size          = 64
        start             = true
        sysctls           = {}
        tmpfs             = {}

      + ulimit {
          + hard = 262144
          + name = "nofile"
          + soft = 100000
        }
      - ulimit {
          - hard = 262144 -> null
          - name = "nofile" -> null
          - soft = 200000 -> null
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

```
$ docker inspect foo 
...
            "Ulimits": [
                {
                    "Name": "nofile",
                    "Hard": 262144,
                    "Soft": 200000
                }
            ],
...
```

Reply:
In short, It succeeds to run `DockerClient.ContainerUpdate` with `ulimit` but actually `ulimit` aren't changed.

Reply:
> Let's merge those 2 comments (L693) into one

Do you suggest to merge comments [R692-693](https://github.com/terraform-providers/terraform-provider-docker/pull/236/files#diff-845dc9c9b3f9ecda8495522e82a0fe8dR691-R692) and [R707-708](https://github.com/terraform-providers/terraform-provider-docker/pull/236/files#diff-845dc9c9b3f9ecda8495522e82a0fe8dR707-R708) into one?
If so, I agree.

Reply:
Update the comment https://github.com/terraform-providers/terraform-provider-docker/pull/236/commits/0f77c1310ac03a5750930a039a41b43e54690a13